### PR TITLE
Reimplement load to be more useable and consistent

### LINKED ISSUE #79: load: allow loading relative paths of arbitrary extension
I'm not sure what the goal of expecting a helper script to be an extension of `.bash` is, but it makes absolutely zero sense.

Let me specify a path and load that file. Don't force the extension on me.

### PR DESCRIPTION
**Pull request by dimo414**
The existing behavior of `load` accepts absolute paths or relative paths - without a `.bash` suffix. This is inconsistent and confusing, both because the `.bash` suffix is arbitrary but also because absolute paths don't have the same constraint.

`load` is re-implemented to behave as a more consistent shorthand for `source`-ing files relative to `BATS_TEST_DIRNAME`, while falling back to `source` the argument as-is if the relative path does not resolve. The `.bash` suffix behavior is preserved for backwards compatibility, but any relative path is now supported. In addition (since `source` supports this natively) scripts on the system `PATH` can also be `load`-ed.

Fixes #79

---

- [x] I have reviewed the [Contributor Guidelines][contributor].
- [x] I have reviewed the [Code of Conduct][coc] and agree to abide by it

[contributor]: https://github.com/bats-core/bats-core/blob/master/docs/CONTRIBUTING.md
[coc]:         https://github.com/bats-core/bats-core/blob/master/docs/CODE_OF_CONDUCT.md

### CODE REVIEW

File: libexec/bats-core/bats-exec-test [hunk 1]
```diff
@@ -40,24 +40,21 @@ fi
...
-  if [[ "${name:0:1}" == '/' ]]; then
-    filename="${name}"
-  else
-    filename="$BATS_TEST_DIRNAME/${name}.bash"
-  fi
+  local file="${1:?name}"
 
-  if [[ ! -f "$filename" ]]; then
-    printf 'bats: %s does not exist\n' "$filename" >&2
-    exit 1
```
Comment (sublimino) [inline libexec/bats-core/bats-exec-test:55]: Could we re-add this error message please? So users don't notice the change : ]

File: libexec/bats-core/bats-exec-test [hunk 2]
```diff
@@ -40,24 +40,21 @@ fi
 BATS_TEST_DIRNAME="${BATS_TEST_FILENAME%/*}"
 BATS_TEST_NAMES=()
 
+# Shorthand for source-ing files relative to the BATS_TEST_DIRNAME,
+# optionally with a .bash suffix appended. If the argument doesn't
+# resolve relative to BATS_TEST_DIRNAME it is sourced as-is.
 load() {
-  local name="$1"
```
Comment (sublimino) [inline libexec/bats-core/bats-exec-test:44]: Apologies for my naivety  — I usually use `:-`, can you explain the behaviour here for me please?

File: libexec/bats-core/bats-exec-test [hunk 3]
```diff
@@ -40,24 +40,21 @@ fi
...
+    file="${BATS_TEST_DIRNAME}/${file}"
+  elif [[ -f "${BATS_TEST_DIRNAME}/${file}.bash" ]]; then
+    file="${BATS_TEST_DIRNAME}/${file}.bash"
   fi
 
-  # Dynamically loaded user files provided outside of Bats.
+  # Dynamically loaded user file provided outside of BATS.
   # shellcheck disable=SC1090
-  source "${filename}"
+  source "${file}" || exit
```
Comment (sublimino) [inline libexec/bats-core/bats-exec-test:57]: What was the previous behaviour here, silent fail? If so perhaps this could be something like

```bash
if ! source "${file}"; then
  printf error message
fi
```

**Comment by dimo414** [comment]
Thanks for the quick review! Looks like there's some existing tests that are affected, I'll tweak them and update the PR.

File: libexec/bats-core/bats-exec-test (hunk 3 above)
Comment (dimo414) [reply libexec/bats-core/bats-exec-test:57]: It would return a non-zero exit code (and print an error), but allow the script to proceed. I imagine the prior assumption was that sourcing would always succeed if the file existed, which isn't necessarily true (e.g. lacking read permission or syntax errors in the file).

That said, this points out another behavior change which may be important. `source` will also return a non-zero exit code if the sourced script's final command does so. If that's a problem we can't actually rely on `source` for error-handling after all (though I doubt users _intend_ for sourced scripts to have non-zero exit codes).

File: libexec/bats-core/bats-exec-test (hunk 2 above)
Comment (dimo414) [reply libexec/bats-core/bats-exec-test:44]: `:?` fails the script if the variable is "null or not set", so this just ensures the user has in fact passed an argument to `load`.

File: libexec/bats-core/bats-exec-test (hunk 1 above)
Comment (dimo414) [reply libexec/bats-core/bats-exec-test:55]: Do we actually expect any users to notice this message change? 😄 You'd typically only see it immediately before your test `exit`s due to a user error (such as assuming relative paths work...), I'm not sure anyone will miss this message. `source` prints an error when it fails to source a file.

That said, the motivation for removing it is to allow `source` to access files on the `PATH`, which `-f` precludes. I believe [something like](https://stackoverflow.com/a/6569837/113632):

    if [[ ! -e "$file" ]] && ! type -P "$file" >/dev/null; then

would suffice, but I was hoping to simplify the implementation by just deferring error-handling to `source`.

Comment (dimo414) [reply libexec/bats-core/bats-exec-test:55]: It looks like the decision was made for us :) [Bash 3.2 seems unhappy with the `|| exit` syntax](https://travis-ci.org/github/bats-core/bats-core/jobs/678477480) (though I'm not exactly sure why), so I went with the approach mentioned above, which preserves the error message.

File: libexec/bats-core/bats-exec-test (hunk 2 above)
Comment (sublimino) [reply libexec/bats-core/bats-exec-test:44]: What's the significance of the `name` value in here, a default if `$1` is an empty string?

File: libexec/bats-core/bats-exec-test [hunk 4]
```diff
@@ -40,24 +40,27 @@ fi
...
-  local filename
+  local file="${1:?name}"
 
-  if [[ "${name:0:1}" == '/' ]]; then
-    filename="${name}"
-  else
-    filename="$BATS_TEST_DIRNAME/${name}.bash"
+  if [[ -f "${BATS_TEST_DIRNAME}/${file}" ]]; then
+    file="${BATS_TEST_DIRNAME}/${file}"
+  elif [[ -f "${BATS_TEST_DIRNAME}/${file}.bash" ]]; then
```
Comment (sublimino) [inline libexec/bats-core/bats-exec-test:51]: I'm wondering now if preserving existing behaviour means searching for the `.bash` version first. In the (perhaps unlikely!) event of both `my_include` and `my_include.bash` existing, would the current implementation default to `my_include` instead of the previous behaviour of `my_include.bash`? Might be easiest-verified with a dedicated test case.

Comment (dimo414) [reply libexec/bats-core/bats-exec-test:51]: I've added a test for the ambiguous behavior, thanks for the suggestion. Given that this change will be held until the 1.3 release how do you feel about keeping the current proposed ordering (exact-match is preferred to `.bash`-suffixed)? If `.bash`-suffixed is preferred `ambiguous` is not reachable, whereas exact-match allows both `ambiguous` and `ambiguous.bash` to be loaded.

If you'd prefer to keep `.bash`-suffix prioritized LMK and I'll flip it and update the test.

---

WRT `${1:?name}`, it is the error that gets reported, and looking at it again I think the default error message is preferable so I've removed `name`. You can experiment with this syntax, try `echo ${foo:?}`, see also [Shell Parameter Expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html).

File: libexec/bats-core/bats-exec-test (hunk 2 above)
Comment (dimo414) [reply libexec/bats-core/bats-exec-test:44]: The text after `:?` is the error that gets reported, and looking at it again I think the default error message is preferable so I've removed it. You can experiment with this syntax, try `echo ${foo:?}`, see also [Shell Parameter Expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html).

File: libexec/bats-core/bats-exec-test (hunk 4 above)
Comment (sublimino) [reply libexec/bats-core/bats-exec-test:51]: I would prefer the `.bash`-suffixed please, it's an edge case and a minor nit but I worry about breaking any backwards compatibility :pray: Then we should be good to go : ]

Comment (dimo414) [reply libexec/bats-core/bats-exec-test]: Sounds good; I left a TODO to consider revisiting this in the future (e.g. if there's a major release).

**Comment by sublimino** [comment]
Thanks for turning this around so quickly @dimo414! :tada:

**Comment by jasonkarns** [comment]
Were docs updated to match new behavior?

**Comment by sublimino** [comment]
@jasonkarns good catch, https://github.com/bats-core/bats-core#load-share-common-code isn't updated.

https://github.com/bats-core/bats-core/blob/master/docs/CHANGELOG.md was updated
